### PR TITLE
Add Admittance BC

### PR DESCRIPTION
**Pull request by jh66637**
Depends on https://github.com/exadg/exadg/pull/631.

This PR also depends on https://github.com/exadg/exadg/pull/619. This dependency is only because the test includes the printed parameter "rhs". I will have to force push to trigger the pipeline once both other PRs are merged. Other than that I am ready here.

In acoustics, admittance boundary conditions can be used to mimic absorbing materials at the domain boundary. We prescribe the BC as follows:
```math
\rho\mathbf{u}\cdot\mathbf{n}=\frac{Y}{c}p
```
Setting $`Y=1`$ corresponds to a first-order absorbing BC (ABC). $`Y=0`$ is equivalent to sound-hard (perfectly reflecting) materials.

### CODE REVIEW

**Comment by nfehn** [comment]
You mention the dependency on other PRs in the description above. Is this resolved now?

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h [hunk 1]
```diff
@@ -65,27 +66,45 @@ template<int dim, typename Number>
...
+      time);
+
+    auto const n      = velocity_m.normal_vector(q);
+    auto const rho_um = velocity_m.get_value(q);
+    auto const pm     = pressure_m.get_value(q);
+
+    auto const rho_um_normal     = (rho_um * n) * n;
+    auto const rho_um_tangential = rho_um - rho_um_normal;
+
+    // normal share of velocty
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:103]: here is a typo: -> velocity

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h [hunk 2]
```diff
@@ -46,13 +46,14 @@ inline DEAL_II_ALWAYS_INLINE //
...
     auto const g = FunctionEvaluator<0, dim, Number>::value(
       *boundary_descriptor.pressure_dbc.find(boundary_id)->second,
-      integrator_m.quadrature_point(q),
+      pressure_m.quadrature_point(q),
       time);
-    return -integrator_m.get_value(q) + Number{2.0} * g;
+    return -pressure_m.get_value(q) + Number{2.0} * g;
   }
-  else if(boundary_type == BoundaryType::VelocityDirichlet)
+  else if(boundary_type == BoundaryType::VelocityDirichlet ||
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:53]: -> `or`

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h [hunk 3]
```diff
@@ -65,27 +66,45 @@ template<int dim, typename Number>
 inline DEAL_II_ALWAYS_INLINE //
   dealii::Tensor<1, dim, dealii::VectorizedArray<Number>>
   calculate_exterior_value_velocity(unsigned int const                       q,
-                                    FaceIntegrator<dim, dim, Number> const & integrator_m,
+                                    FaceIntegrator<dim, dim, Number> const & velocity_m,
+                                    FaceIntegrator<dim, 1, Number> const &   pressure_m,
+                                    double const                             c,
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:71]: We should use here `speed_of_sound` as is used below for the `BoundaryFaceIntegrator`.

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h [hunk 4]
```diff
@@ -65,27 +66,45 @@ template<int dim, typename Number>
...
+
+    auto const n      = velocity_m.normal_vector(q);
+    auto const rho_um = velocity_m.get_value(q);
+    auto const pm     = pressure_m.get_value(q);
+
+    auto const rho_um_normal     = (rho_um * n) * n;
+    auto const rho_um_tangential = rho_um - rho_um_normal;
+
+    // normal share of velocty
+    auto const rho_u_n = (Number{2.0} * Y * pm / c) * n - rho_um_normal;
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:104]: To have a consistent notation, this variable should read `rho_up_normal` if I understand the notation correctly that is used for the other variables?

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h [hunk 5]
```diff
@@ -134,13 +153,19 @@ template<int dim, typename Number>
 class BoundaryFaceIntegratorU : public BoundaryFaceIntegratorBase<BoundaryDescriptor<dim>, Number>
 {
   using FaceIntegratorU = FaceIntegrator<dim, dim, Number>;
+  using FaceIntegratorP = FaceIntegrator<dim, 1, Number>;
 
 public:
-  BoundaryFaceIntegratorU(FaceIntegratorU const &         integrator_m_in,
+  BoundaryFaceIntegratorU(FaceIntegratorU const &         integrator_velocity_m_in,
+                          FaceIntegratorP const &         integrator_pressure_m_in,
+                          double const                    speed_of_sound,
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:161]: From a pure software design perspective, the dependency on `speed_of_sound` could be seen detrimental. One technical solution to this "problem" could be to integrate the factor `c` into the `dealii::Function` that we already use to prescribe `Y`.

File: include/exadg/acoustic_conservation_equations/user_interface/boundary_descriptor.h [hunk 6]
```diff
@@ -44,13 +44,17 @@ namespace Acoustics
  * | prescribe velocity values |                         | Dirichlet:              |
  * |                           | no BCs to be prescribed | prescribe g_u           |
  * +---------------------------+-------------------------+-------------------------+
+ * | admittance BC             |                         | Admittance:             |
+ * |                           | no BCs to be prescribed | prescribe Y             |
+ * +---------------------------+-------------------------+-------------------------+
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/user_interface/boundary_descriptor.h:49]: I think somewhere here the equation provided in the description of this PR needs to be written for documentation. Otherwise, it is not clear to the user where the factor `rho` and `c` appear and whether `Y` already accounts for them or not. Please also note comment https://github.com/exadg/exadg/pull/629/files#r1450469079 in this context.

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 5 above)
Comment (nfehn) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:161]: What do you think?

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h [hunk 7]
```diff
@@ -65,27 +66,45 @@ template<int dim, typename Number>
...
+    auto const rho_um = velocity_m.get_value(q);
+    auto const pm     = pressure_m.get_value(q);
+
+    auto const rho_um_normal     = (rho_um * n) * n;
+    auto const rho_um_tangential = rho_um - rho_um_normal;
+
+    // normal share of velocty
+    auto const rho_u_n = (Number{2.0} * Y * pm / c) * n - rho_um_normal;
+
+    // tangential share of velocity stays the same
```
Comment (nfehn) [inline include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:106]: For clarity, I would add here "(i.e. rho_up_tangential = rho_um_tangential)".

File: applications/acoustic_conservation_equations/normal_incidenting_wave/application.h [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    {
+      // To be able to compute an error and thus test the behavior of the admittance BC for
+      // Y=1 (first-oder ABC). We also have to ensure the that the wave left the domain
+      // when computing the error.
+      AssertThrow(std::abs(admittance - 1.0) < 1.0e-12,
+                  dealii::ExcMessage("The analytical solution only holds for Y=1."));
+      AssertThrow(this->param.end_time > 1.5 * radius / speed_of_sound,
+                  dealii::ExcMessage("The end time has to be after the wave left the domain."));
+
+      // calculation of pressure error
```
Comment (nfehn) [inline applications/acoustic_conservation_equations/normal_incidenting_wave/application.h:205]: Velocity does not make sense here?

File: applications/acoustic_conservation_equations/normal_incidenting_wave/application.h [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+                        "Admittance.",
+                        dealii::Patterns::Double(0.0, 1.0));
+
+      prm.add_parameter("EndTime", end_time, "End time.", dealii::Patterns::Double(1.0e-12));
+
+      prm.add_parameter("Radius", radius, "Radius of domain.", dealii::Patterns::Double(1.0e-12));
+    }
+    prm.leave_subsection();
+
+    // For a reasonable test we have to create a differen postprocessor, see below.
```
Comment (nfehn) [inline applications/acoustic_conservation_equations/normal_incidenting_wave/application.h:88]: typo ->different

File: applications/acoustic_conservation_equations/normal_incidenting_wave/application.h [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    prm.enter_subsection("General");
+    {
+      prm.add_parameter("IsTest", is_test, "Do we run a test?", dealii::Patterns::Bool());
+    }
+    prm.leave_subsection();
```
Comment (nfehn) [inline applications/acoustic_conservation_equations/normal_incidenting_wave/application.h:93]: I am not sure whether we should use the existing variable `is_test` for this. I think I would go for an application-specific variable like `analytical_solution_is_available` or `use_first_order_admittance_bc`.

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 5 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:161]: From a pure software perspective, this is true. I would still like to keep it this way because in literature values for admittance or impedance ($1/Y$) are provided, and it is not intuitive to prescribe $Y/c$ from a physical acoustic perspective.

File: applications/acoustic_conservation_equations/normal_incidenting_wave/application.h (hunk 10 above)
Comment (jh66637) [reply applications/acoustic_conservation_equations/normal_incidenting_wave/application.h]: [ce3f9a4](https://github.com/exadg/exadg/pull/629/commits/ce3f9a4cff7cfd9db600da22d60e3046c461f9e7)

File: applications/acoustic_conservation_equations/normal_incidenting_wave/application.h (hunk 9 above)
Comment (jh66637) [reply applications/acoustic_conservation_equations/normal_incidenting_wave/application.h]: [4a39a95](https://github.com/exadg/exadg/pull/629/commits/4a39a95edf561b170486d898f80d5e7117c9a749)

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 1 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h]: [b1096ed](https://github.com/exadg/exadg/pull/629/commits/b1096ed3c7a3a0caa43ef4c87ffb4cc5ab7f8a3d)

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 3 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h]: [4b4c61d](https://github.com/exadg/exadg/pull/629/commits/4b4c61dfcfc1ab1fe9137668b6ec6d09ab57e672)

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 4 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h]: [e17dfc4](https://github.com/exadg/exadg/pull/629/commits/e17dfc41bddbcfd1dc4dcfe79539acfc61e10bec)

File: applications/acoustic_conservation_equations/normal_incidenting_wave/application.h (hunk 8 above)
Comment (jh66637) [reply applications/acoustic_conservation_equations/normal_incidenting_wave/application.h:200]: [f6db7a7](https://github.com/exadg/exadg/pull/629/commits/f6db7a71b09ace5010d6443011cf38242c4702bb)

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 7 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h]: [af66293](https://github.com/exadg/exadg/pull/629/commits/af6629314c29799f74213d1444ae367f1ac49c2f)

File: include/exadg/acoustic_conservation_equations/user_interface/boundary_descriptor.h (hunk 6 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/user_interface/boundary_descriptor.h]: [b81837d](https://github.com/exadg/exadg/pull/629/commits/b81837d0c9fa450f3a3ed37d6051e77712891b91)

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 2 above)
Comment (jh66637) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h]: [403fb50](https://github.com/exadg/exadg/pull/629/commits/403fb50cd7906d6839c08ff2b130489766a3a6bb)

**Comment by jh66637** [comment]
@nfehn Thank you for the review. I rebased and there is no dependency on other PRs anymore. I also addressed all your comments.

File: include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h (hunk 5 above)
Comment (nfehn) [reply include/exadg/acoustic_conservation_equations/spatial_discretization/operators/weak_boundary_conditions.h:161]: I'm fine with it for now. I take this as an indication that generalization `BoundaryFaceIntegrator`s across modules (acoustics, Poisson, compressible) is not reasonable (for now) for reasons that we also encounter here.

**Comment by nfehn** [comment]
@jh66637 thanks for the updates!
